File: target_motion.m
function [target, c_sphere] = target_motion(target, p_d, Param)
%TARGET_MOTION Propagates target drone one timestep
%
% Inputs:
%   target   - struct with fields x (18x1) and xiep (3x1)
%   p_d      - desired position (3x1)
%   Param    - parameter struct
%
% Outputs:
%   target   - updated target struct
%   c_sphere - target position (3x1)

    % Extract state
    pos = target.x(1:3);
    v   = target.x(4:6);
    R   = reshape(target.x(7:15),3,3);
    om  = target.x(16:18);

    % Desired quantities (simple hover / trajectory tracking)
    v_d    = zeros(3,1);
    a_d    = zeros(3,1);
    j_d    = zeros(3,1);
    psi_d  = 0;
    dpsi_d = 0;

    % Controller (reuse drone controller)
    [T, tau, e_p] = drone_mellinger_ctrl( ...
        pos, v, R, om, Param, ...
        p_d, psi_d, target.xiep, ...
        v_d, dpsi_d, a_d, j_d);

    % Integrate position error
    target.xiep = target.xiep + Param.dTi*e_p;

    % Drone dynamics
    [dot_p, dot_v, dot_R, dot_om] = ...
        drone_3dfull_dyn(v, R, om, T, tau, Param);

    % Discretization (same as your agents)
    pos_p = pos + Param.dTi*dot_p*0.5;
    v_p   = v   + Param.dTi*dot_v*0.5;
    R_p   = rot_integrate(R, om, Param.dTi);
    om_p  = om  + Param.dTi*dot_om;

    % Pack state
    target.x = [ ...
        pos_p;
        v_p;
        reshape(R_p,[],1);
        om_p ];

    % Output target position
    c_sphere = pos_p;
end

File: Python/target_motion.py
import numpy as np
from Structs import DroneState, Param
from drone_3dfull_dyn import drone_3dfull_dyn
from helper_funcs import *
from drone_mellinger_ctrl import drone_mellinger_ctrl

def target_motion(target: DroneState, param: Param):
    """
    Propagates the target drone one timestep.

    Parameters
    ----------
    target : dict
        Target state with keys:
            - "pos" : (3,1)
            - "vel" : (3,1)
            - "attitude" : (3,3) rotation matrix
            - "ang_vel" : (3,1)
    p_d : ndarray
        Desired position (3,1)
    v_d : ndarray
        Desired velocity (3,1)
    Param : dict
        Simulation parameters.
    """

    # Extract state
    pos = target.position
    vel = target.velocity
    R = target.attitude.reshape(3,3)
    om = target.ang_vel

    # Desired quantities (trajectory tracking)
    dpsi_d = (target.desired.psi - target.desired.old_psi)/param.sim.dTi
    target.desired.old_psi = target.desired.psi

    # Controller
    thrust, torque = drone_mellinger_ctrl(
        target,
        dpsi_d, 
        param
    )

    # Drone dynamics
    dot_p, dot_v, dot_om = drone_3dfull_dyn(
        vel,
        R,
        om,
        thrust,
        torque,
        param,
    )

    # Euler integration
    dt = param.sim.dTi

    target.position = pos + dt * dot_p
    target.velocity = vel + dt * dot_v
    target.attitude = rot_integrate(R, om, dt).reshape(-1,1)
    target.ang_vel = om + dt * dot_om